kernel/work_q.c

```diff
@@ -99,7 +99,9 @@ int k_delayed_work_submit_to_queue(struct k_work_q *work_q,
 		/* -EALREADY indicates the work has already completed so this
 		 * is likely a recurring work.
 		 */
```

Comment:
This seems like further circumvention of the problem to me.
It would be better to have a better check for if the work should be canceled or not. because in this case it doesn't need to be canceled, which is why it returns EALREADY.
The work->work_q is not reset to NULL once finished, so it should be used in all this error handling.

The check above for example:
```
	if (work->work_q != NULL && work->work_q != work_q) {
		err = -EADDRINUSE;
		goto done;
	}
```
Is this one the correct behavior?

Reply:
I would pick https://github.com/zephyrproject-rtos/zephyr/issues/22933#issuecomment-588287754 as the solution.

Reply:
That seems like something that should go into work_cancel(), not here, though, right?

Reply:
Yep, I though about moving that check ou of work_cancel now that you mentioned it might be the right thing to do as a separate patch.